### ISSUE TITLE
Better support for following pixelfed users.

### ISSUE THREAD

**Issue by kr428** [mastodon/mastodon#18127]
### Pitch

Currently, though I can follow pixelfed users from Mastodon, this feature is somewhat "crippled":

* pixelfed allows more pictures per post than Mastodon, apparently these additional images simply aren't displayed without any notice.
* pixelfed supports stories which don't appear at all in Mastodon and I have no way to even see the user makes use of these.
* pixelfed supports licenses (CC-BY-SA) and location attached to image posts, which isn't visible at all while looking at pixelfed posts from a Mastodon user interface.

Please consider ways to make this information visible in the Mastodon user interfaces, too.

### Motivation

Currently, the ability to subscribe and follow users on other platforms is one of the _big_ arguments for pulling people over to Mastodon and the fediverse. It's minor nuissances and inconveniences like these that really hinder this adoption and are likely to turn users away (once they learn that, though they _can_ follow pixelfed users from within mastodon, they will lose a considerable amount of information this way).

**Comment by nightpool** [comment]
Of these, I think the only one that makes sense to add is a "more pictures available" link that could maybe take you back to the original image. Extra metadata like license and location aren't really important to show on remote instances, you can always go back to the original instance to see them.

Stories would be nice, but it would require Mastodon to implement the entire stories feature, there's no way to "translate" stories into a Mastodon-friendly format without changing the UX expectations (e.g. ephemerality and separation from normal post content).

**Comment by kr428** [comment]
> Of these, I think the only one that makes sense to add is a "more pictures available" link that could maybe take you back to the original image. Extra metadata like license and location aren't really important to show on remote instances, you can always go back to the original instance to see them.
> 

I get your point but see two "issues" here:

- First and foremost, with the idea of visiting remote instances to see full data, users essentially will have to do so every time in order to make sure they're not missing something as (for stories) within Mastodon they don't even have a chance of seeing there's information they don't have access to as "followers" from wherever they are. 
- Worse: This approach doesn't at all seem to work for private profiles and content. In this case, navigating to the web instance will require users to (a) have a pixelfed account to log in with and (b) be following the same users from that account too. 

So at this point, it would mean falling back to the walled garden approach of requiring separate accounts per platform again, anyhow, or maybe it will be even more difficult to handle: For Twitter vs. Instagram, people *know* these accounts aren't capable of interacting with each other, as these are proprietary silos where exchange and interactions across these borders isn't desired. Exactly the opposite is one of the core salespoints of ActivityPub and the federated networks, and it also works - but just to some degree, and for users it is hard to impossible to figure out to which degree and what this kind of compatibility actually depends upon... 😔️

**Comment by d4f5409d** [comment]
The license and location support would be amazing.

**Comment by wolfspyre** [comment]
interop is def a little clunky… 

what I would LOVE to see is a better way to deal with posts that have more than 4 pieces of media content …

what can help amplify this/increase the priority of it?


additionally being able to elect not to cache / duplicate content from. an endpoint/server would be helpful (for example, when one is hosting a masto instance alongside a pixelfed or peertube instance

**Comment by kr428** [comment]
> interop is def a little clunky…
> 
> what I would LOVE to see is a better way to deal with posts that have more than 4 pieces of media content …
> 
> what can help amplify this/increase the priority of it?
> 

Slightly off here but as I'm the author of this issue, I have to say I have these days given up on both Mastodon and pixelfed for particular reasons. And am very much doubting current ActivityPub specs in general. Because, after all ... this /still/ is an issue but so far it seems just the tip of an iceberg of flawed interoperability between various platforms, and (hate to say that but still) it seems Mastodon is both the still "biggest" crowd of AP users out there /and/ the platform that is most hesitant when it comes to improving interoperability with other implementations. Maybe the best way to increase priority here is (also) using other platforms and move these interoperability aspects from something that can easily be ignored to something that needs to be addressed because it also affects Mastodon users directly. Or put more focus and energy in or on the FEP proposals / project - https://codeberg.org/fediverse/fep - to make interoperability a more crucial core part of the protocol itself, although I am not overly optimistic here.

**Comment by sjlongland** [comment]
> Slightly off here but as I'm the author of this issue, I have to say I have these days given up on both Mastodon and pixelfed for particular reasons. And am very much doubting current ActivityPub specs in general. Because, after all ... this /still/ is an issue but so far it seems just the tip of an iceberg of flawed interoperability between various platforms, and (hate to say that but still) it seems Mastodon is both the still "biggest" crowd of AP users out there /and/ the platform that is most hesitant when it comes to improving interoperability with other implementations.

It actually reminds me a lot of the browser wars of the 90s.

You had two major players in the beginning, NCSA Mosaic and Netscape.  Netscape had all the flashy features (frames, JavaScript, plug-ins, `<blink>` tag), but Mosaic and all its clones (including Microsoft Internet Explorer v1 and v2) offered basic functionality… then Microsoft really got stuck in with v3 of their browser: they added special features like ActiveX controls, the `<marquee>` tag, rudimentary CSS.  Netscape added similar level rudimentary CSS in v4 of their browser whilst Microsoft kept chasing the bleeding edge.

You wound up with a situation where the same site displayed wildly different in the two major browsers.  Then additional players like Opera entered the scene with their take on standards.

It was a mess, made worse when the then dominant player decided they were "done" (Microsoft with the release of IE6) and sat on their laurels.  It took Apple forking KHTML (from the KDE project) to produce Webkit, and Google picking said Webkit up to produce Chrome (eventually forking Webkit for Blink), to finally unseat the situation.

Now, will that eventually happen with ActivityPub?  Maybe.  We've got an advantage in that most of our implementations here are open-source community-run projects, not commercial proprietary implementations.  (I think there's Threads and Medium, with Imgur showing interest.)  It'll still take time though, so we need patience.

**Comment by kr428** [comment]
> 
> Now, will that eventually happen with ActivityPub? Maybe. We've got an advantage in that most of our implementations here are open-source community-run projects, not commercial proprietary implementations. (I think there's Threads and Medium, with Imgur showing interest.) It'll still take time though, so we need patience.

There's a difference in between the things I /hope/ and the things I /see/ here. Hope-wise, I'm definitely in your boat. Realistically speaking, this is slightly different and issues like this somehow seem to speak more for that second strain of thoughts. I have not much to add except for linking to this article ... https://overengineer.dev/blog/2019/01/13/activitypub-final-thoughts-one-year-later/ .. by one of the Diaspora guys, posted years ago on AP standardization and its ... flaws also when it comes to usability. At some point he came to this conclusion...

> When I wrote the article last year, I honestly thought everyone was building federated systems for their users, not for their warm feeling for implementing some new standards. I was wrong, and people have now accepted that whatever comes out at the end is probably not a good solution for their users, and whenever this point comes up, the critique simply gets disregarded. This is deliberately throwing non-technical user’s expectations under the bus.

... which he used to write in early 2019, and, now, six years and something later, I still feel like wrestling the same usability glitches in the Fediverse that used to be around back then, plus a bunch of others arising from even more different platforms being around providing different, slightly incompatible implementations of the standard. It feels the AP  spec, even decades later, hasn't  learnt much from what caused the browser wars and should probably not have been repeated. But well here we are.

**Comment by sjlongland** [comment]
Yeah, "prestige" isn't a good reason to write a spec.  I admire what ActivityPub's spec has achieved, and still remain hopeful, but see the specification stagnate over politics is not great.  …and politics is an inevitable consequence of a large open project with lots of stakeholders.

Unfortunately the old saying still rings true.  History never repeats, but it often rhymes.

**Comment by wolfspyre** [comment]
look at us .... a bunch of weathered experienced somewhat crusty sweethearts... saddened that the pains we experienced decades ago are remanifesting.... 

This is my concern too ... and my intent is to ... well ... show up with punch and pie ... or come to the party with beer .... rather than point at a thing and say 'brokey! noworky!' ....  

I want to help this go in the least wrong direction ... the struggle of course is 
> By who's lens?

**MY** personal desire, is to provide a means for my community to share content via a means that is low-effort, and have their friends/fams/peers/followers have an easy and enjoyable time consuming their content via a user experience that doesn't suck .... 

The challenge here is everyones' need is different, and solutions need to be adaptable to fit those divergent paths ... 

to a large extent, this is why the dif ap apps exist ....  Also worth mentioning ... Managing this stuff isn't easy for almost anyone ...  Complex systems are .. well ... by their nature... complex :) 

Not everyone running their own (insert fediverse endpoint here) has the same technical expertise or drive .... and that's okay ... (mostly)...  but it also cannot be the primary driver, because every choice has consequences.... 
overoptimizing for 'ease of administration' results in narrow  applicability and poor UX... 

Then again .... designing large systems that interact in complex ways is **HARD** and requires balancing many things... which requires a collectively aligned vision of where everyone's trying to go .... 
 
Give people enough papercuts and their willingness to continue to endure them starts to drastically wane .... 

I have immense gratitude for the developers and kind souls who are architecting and contributing to these systems... 

**AND** I feel somewhat powerless to help move things towards alignment... 

I don't want to sound like I'm griping about things I don't understand, nor like I'm being flippant about others hard work... or that I don't understand the complex balancing act that managing foss projects at scale represents .... 

LOTS of the peeps who've weighed in on various issues have been participating in the internet for decades ... and have the scars from lived experience .... 

AND we all only have so many f's to give .... and the world around us has kinda taken way more of them lately than I think ANY of us want...  corespondingly we're all...... a lil rough around the edges these days ... 

which makes collaborative projects EXTRA hard .... 

I'm seeing this manifest across multiple foss project contexts ... it's not just this bubble .... it's .... EVERYWHERE .... I cannot point to one cohort of peeps I know that is thriving right now .... 

SO.... keeping in mind that most of us are NOT our best selves right now ... 
AND that it's stupendously easy to misinterpret intention from text, as its SUCH a lossy medium .... 

I'd like to reiterate that AP / Fediverse interop is hard, important, and has to be prioritized.... which I am fairly sure that it largely is... but we all have facets that are more or less important to us because reasons.... 

I would like to see this aspect of actor experience within masto improve... 

What would help that happen?


❤️🐺W

**Comment by wolfspyre** [comment]
> Of these, I think the only one that makes sense to add is a "more pictures available" link that could maybe take you back to the original image. Extra metadata like license and location aren't really important to show on remote instances, you can always go back to the original instance to see them.
> 
> Stories would be nice, but it would require Mastodon to implement the entire stories feature, there's no way to "translate" stories into a Mastodon-friendly format without changing the UX expectations (e.g. ephemerality and separation from normal post content).

I disagree with this lens, and I think harboring it does a disservice to the mastodon userbase ... 

I'll articulate my reasoning. Please feel INVITED to help me see what I'm (clearly) missing ...  


I'd think it reasonable to assert that any federatable event aught be visible in any client, to some extent...  

that's what being a member of a federated system means, in my mind .... pubsubby I push a thing out and consumers consume and present them in a manner that's contextually proper... 

an ephemeral event is still an event... just one that will end, and the content that was there may or may not be retained after the fact... but the event did still occur, and as such aught be representable within a timeline of 'stuff that occurred' .... 

HOW that's represented of course, will vary immensely depending on context n wotnot ... 

I can easily see the stance that 
> 'its the discretion of the actor as to whether or not they wish to see record of expired ephemeral events in a given timeline ... ' 

I can also see the stance that 
> content authors should have the ability to declare durability of content, and have that respected. 

As it sits now, I'm struggling with trying to identify the "right" way forward, in the context of: 

My pack of artists wish to have the ability to share with their users pictures of the stuff they make ... in a way that is easy for their followers to consume.... 

that's reasonable... I want the end user to be able to easily discover content, and be able to see it all... not just some of it.... Currently, the masto user is not made aware that something is being obfuscated, 

Thus, my perspective is that Masto's current behavior is broken. 

I'm not saying that masto must show everything.... 

(doing so would likely be somewhat problematic in many contexts) 
HOWEVER, choosing to prevent masto actors from even **KNOWING THAT THEY ARE MISSING CONTENT** 

.... 

That's what's broken. 
Before I start pushing for a more specific change than 'the user should be aware that content exists (or existed) that they cannot contextually see all of from the masto UI they are using at the moment' .... 

I'd like to understand why this behavior exists... I'm sure there was (and perhaps still is?) a good reason for the choices that were made ... 

I don't know if that reason is still relevant or not..... however in my lens the end user deserves to be informed that they are unable to see all the content (or all the attributes of said content) that would otherwise be visible to them... and have some (not egregiously difficult) path to do so.... 

That feels like a reasonable goal that would benefit all activitypub actors equally .... 


My intent here ain't to sling mud or anything.... I'm approaching this from the lens of 

'what is the best experience for the end user?' 

I am unaware of any reason from a technical perspective that this shouldn't be attainable within the current tools... 

y'all have been thinking about this, and immersed in this space WAY LONGER than I have.....  

What am I missing?

**Comment by kr428** [comment]
> (doing so would likely be somewhat problematic in many contexts)
> HOWEVER, choosing to prevent masto actors from even **KNOWING THAT THEY ARE MISSING CONTENT**

Exactly _this_ is my biggest pain point with this situation. Being an artist on pixelfed, it's annoying that some of your pictures aren't to be seen by Mastodon users without them even getting a hint that this might be a problem. If you're, like, a journalist creating a photo-backed report on some critical issue maybe on pixelfed, having just displayed four out of ten pictures isn't even half the story. And, even worse, as a creator too I do not necessarily know that platforms other than mine decide to willingly and quietly discard information. My initial intention was to come up with a feature request here, but the more I think about it, the more this looks like a critical bug resulting in loss of information.

(And, looking at it from another way, it resembles the age of browser wars but even worse. Remember building and testing websites on all major browsers? By now it seems to even more boil down to having accounts on at least one instance of all  AP services to test and be sure your posts over there behave how you intended them to.)

**Comment by wolfspyre** [comment]
I do really, truly, completely want the decentralized social amalgam to thrive.

I believe this effort is CRUCIAL to maintaining our ability to collectively share with, inspire, and learn from others in ways that help everyone…

the large platforms which have the resources, thanks to selling your attention at least…. your data at worst..  have a lot more control over what their userbase sees… 

this ecosystem IS THE ANTIDOTE TO THAT …

(even as kinda klugey as it is, it’s better than (insert forprofitcorp here) )


please please please….. 
for the love of beauty, porn, crazy cat people, freaky furries, and all the rest of us marginalized weirdos ….

can we find a way forward that doesn’t suck unnecessarily?


I want to understand the reasoning behind these pain points…. 

if the reasoning is no longer relevant, can we try and prioritize interop a bit more vigorously?

I know we’re all sick of the enshittification proliferating around us…

can we try not to perpetuate it? :) ❤️🐺w

**Comment by wolfspyre** [comment]
related:
- https://github.com/mastodon/mastodon/issues/10258 
- https://github.com/mastodon/mastodon/issues/14336
